File: frontend/src/utils/cronCommand.ts
import { hasControlCharacter, hasEdgeWhitespace } from './cronText'

/**
 * The client-side mirror of the panel's `CronCommandRule`: what this SPA accepts as one cron
 * command line.
 *
 * **A mirror, and only a mirror.** The module re-validates the command and the agent re-validates
 * it again; checking here is what lets an operator be told what is wrong with the line they typed
 * instead of being handed a refusal from a process they cannot see (rules/vue.md).
 *
 * The alphabet is a short list of refusals rather than a permitted set, and that is the module's
 * decision faithfully reproduced. The command never reaches the crontab — the agent writes it to a
 * per-entry file under the account's home and the installed crontab line runs that file — so the
 * two characters a crontab line genuinely cannot carry are ordinary text here. Refusing a percent
 * sign or a hash would refuse working commands for a danger that does not exist at this position.
 */

/**
 * The most bytes a command may be, matching the module's own ceiling.
 *
 * Measured in UTF-8 BYTES, not in UTF-16 code units, because the agent measures the bytes it
 * writes. A command of three thousand emoji is three thousand characters and twelve thousand bytes:
 * counted as characters it would pass here and be refused after the operator had been told their
 * entry was accepted.
 */
const MAXIMUM_LENGTH_IN_BYTES = 4096



/** Measures a string the way the module does — in UTF-8 bytes. */
const encoder = new TextEncoder()

/**
 * Whether a candidate is one acceptable cron command line.
 *
 * Surrounding whitespace is refused rather than trimmed. The command is stored verbatim and
 * compared verbatim when the agent decides whether an entry duplicates one already installed, so a
 * leading space and a trailing space must not become two spellings of one command — and trimming
 * silently would show the operator something other than what they typed.
 * @param candidate The command line as the operator typed it.
 * @returns True when it is a non-empty, bounded, single line with no control character and no
 * leading or trailing whitespace.
 */
export const isOneCronCommandLine = (candidate: string): boolean => {
  if (candidate.length === 0) {
    return false
  }

  if (encoder.encode(candidate).length > MAXIMUM_LENGTH_IN_BYTES) {
    return false
  }

  // What a FILE holding exactly one line cannot carry. Whitespace is not in this class — a command
  // is full of spaces — so the edges are a separate check, exactly as they are in the module.
  if (hasControlCharacter(candidate)) {
    return false
  }

  return !hasEdgeWhitespace(candidate)
}

File: agent/crates/agent-core/src/validation/system/cron_command.rs
//! The command a cron entry runs, checked before it is written to its own file.

use super::cron_command_error::CronCommandError;

/// The most bytes a command may be.
///
/// A `.cmd` file holds ONE command line, not a script: a customer who needs a
/// script writes one and schedules a call to it. Four kilobytes is far more
/// than any such line needs and is small enough that a thousand entries cost
/// nothing to read back.
const MAX_LENGTH: usize = 4096;

/// A validated cron command — one shell command line, stored verbatim.
///
/// The inner string is private and the only constructor is
/// [`CronCommand::parse`], so holding a value of this type is proof that
/// validation happened.
///
/// This value is written to a per-entry file under the account's home and run
/// by `/bin/sh <file>` from a crontab line that contains no byte of it. That
/// placement is the whole security design and it is why the alphabet here is a
/// short list of refusals rather than a permitted character set: the characters
/// a crontab line cannot carry — `%`, which cron rewrites into a newline, and
/// `#`, which starts a comment — are ordinary shell text in a file, and
/// refusing them would refuse working commands for a danger that no longer
/// exists. [`CronCommandError`] carries the full reasoning, including the two
/// commands that disproved the earlier in-line design on a real host.
///
/// What the type still guarantees is that the value is one line: no control
/// characters, no surrounding whitespace, and a bounded length.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct CronCommand(String);

impl CronCommand {
    /// Validates `candidate` as one cron command line and wraps it.
    ///
    /// Accepts 1 to 4096 bytes of UTF-8 — the parameter is a `&str`, so
    /// well-formed UTF-8 is the caller's problem and already solved — with no
    /// control character anywhere and no leading or trailing whitespace.
    /// Everything else is allowed, `%` and `#` included.
    ///
    /// # Errors
    ///
    /// - [`CronCommandError::Empty`] when `candidate` is empty.
    /// - [`CronCommandError::TooLong`] when it exceeds 4096 bytes.
    /// - [`CronCommandError::ControlCharacter`] for a newline, a carriage
    ///   return, a tab, a NUL or any other control character.
    /// - [`CronCommandError::SurroundingWhitespace`] when it begins or ends
    ///   with whitespace.
    pub fn parse(candidate: &str) -> Result<Self, CronCommandError> {
        if candidate.is_empty() {
            return Err(CronCommandError::Empty);
        }

        if candidate.len() > MAX_LENGTH {
            return Err(CronCommandError::TooLong {
                maximum: MAX_LENGTH,
            });
        }

        // Before the whitespace check, so that a newline is reported as the
        // control character it is rather than as a stray blank at an end.
        if let Some(character) = candidate.chars().find(|c| c.is_control()) {
            return Err(CronCommandError::ControlCharacter { character });
        }

        if candidate.starts_with(char::is_whitespace) || candidate.ends_with(char::is_whitespace) {
            return Err(CronCommandError::SurroundingWhitespace);
        }

        Ok(Self(candidate.to_owned()))
    }

    /// The validated command, exactly as it was written.
    #[must_use]
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

#[cfg(test)]
#[path = "../../tests/validation/system/cron_command_tests.rs"]
mod tests;
